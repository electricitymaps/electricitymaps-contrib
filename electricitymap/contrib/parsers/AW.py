#!/usr/bin/env python3

from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

from .lib.exceptions import ParserException

FUEL_MAPPING = {
    "Fossil": "oil",
    "Wind": "wind",
    "TotalSolar": "solar",
    "total_bio_gas": "biomass",
}
TIMEZONE = ZoneInfo("America/Aruba")
PRODUCTION_URL = (
    "https://www.webaruba.com/renewable-energy-dashboard/app/rest/results.json"
)
SOURCE = "webaruba.com"


def fetch_production(
    zone_key: ZoneKey = ZoneKey("AW"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    r = session or Session()
    # User agent is mandatory or services answers 404
    headers = {
        "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)"
    }
    response = r.get(PRODUCTION_URL, headers=headers)
    aruba_json = response.json()
    top_data = aruba_json["dashboard_top_data"]
    mix = ProductionMix()
    sources_total = 0
    for fuel, mode in FUEL_MAPPING.items():
        value = float(top_data[fuel]["value"])
        mix.add_value(mode, value)
        sources_total += value
    # "unknown" is when data reported in the categories above is less than total reported.
    # If categories sum up to more than total, accept the datapoint, but only if it's less than 10% of total.
    # This helps avoid missing data when it's a little bit off, due to rounding or reporting
    reported_total = float(top_data["TotalPower"]["value"])

    if (sources_total / reported_total) > 1.1:
        raise ParserException(
            "AW.py",
            f"AW parser reports fuel sources add up to {sources_total} but total generation {reported_total} is lower",
            zone_key,
        )

    mix.add_value(
        "unknown", reported_total - sources_total, correct_negative_with_zero=True
    )
    # We're using Fossil data to get timestamp in correct time zone
    local_date_time = datetime.strptime(
        top_data["total_bio_gas"]["timestamp"], "%Y-%m-%d %H:%M:%S.%f"
    ).replace(tzinfo=TIMEZONE)

    production_list = ProductionBreakdownList(logger)
    production_list.append(
        zone_key,
        local_date_time,
        SOURCE,
        production=mix,
    )

    return production_list.to_list()


if __name__ == "__main__":
    print(fetch_production())
