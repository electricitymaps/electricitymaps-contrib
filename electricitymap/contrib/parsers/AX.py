from datetime import datetime, timedelta
from logging import Logger, getLogger
from re import findall
from zoneinfo import ZoneInfo

from bs4 import BeautifulSoup
from requests import Response, Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
    TotalConsumptionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

from .lib.exceptions import ParserException

IFRAME_URL = "https://grafik.kraftnat.ax/grafer/tot_inm_24h_15.php"
TIME_ZONE = ZoneInfo("Europe/Mariehamn")
SOURCE = "kraftnat.ax"


def fetch_data(session: Session, logger: Logger):
    """Fetch data from the iFrame."""

    response: Response = session.get(IFRAME_URL)
    response.raise_for_status()

    soup = BeautifulSoup(response.text, "html.parser").find_all("script")
    result_time_series = findall(r"data: \[(.*?)\]", str(soup))
    if len(result_time_series) != 3:
        raise ParserException(
            "AX.py",
            "Did not find the expected amount of results. Check if the website has changed.",
        )
    time_series: list = result_time_series[0].split(",")
    raw_data: list[str] = findall(r"data:\[(.*?)\]", str(soup))
    if len(raw_data) != 6:
        raise ParserException(
            "AX.py",
            "The raw data did not match the expected format. Check if the website has changed.",
        )
    for raw in raw_data:
        if len(raw.split(",")) != len(time_series):
            raise ParserException(
                "AX.py",
                "The raw data did not match the length of the the time series. Check if the website has changed.",
            )
    data_list = []
    for time, sweden, alink, fossil, gustavs, wind, consumption in zip(
        time_series,
        raw_data[0].split(","),
        raw_data[1].split(","),
        raw_data[2].split(","),
        raw_data[3].split(","),
        raw_data[4].split(","),
        raw_data[5].split(","),
        strict=True,
    ):
        data_list.append(
            {
                "time": str(time.replace('"', "")),
                "sweden": float(sweden),
                "alink": float(alink),
                "fossil": float(fossil),
                "gustavs": float(gustavs),
                "wind": float(wind),
                "consumption": float(consumption),
            }
        )
    return data_list


def formatted_data(
    zone_key: ZoneKey | None,
    zone_key1: ZoneKey | None,
    zone_key2: ZoneKey | None,
    session: Session,
    logger: Logger,
    data_type: str,
):
    """Format data to Electricity Map standards."""
    data_list = fetch_data(session, logger)
    data_list.reverse()
    date_time = datetime.now(TIME_ZONE)
    date = date_time.replace(
        hour=int(data_list[0]["time"].split(":")[0]),
        minute=int(data_list[0]["time"].split(":")[1]),
        second=0,
        microsecond=0,
    )
    if date > date_time:
        date = date - timedelta(days=1)
    exchanges = ExchangeList(logger)
    production_breakdowns = ProductionBreakdownList(logger)
    consumption = TotalConsumptionList(logger)
    for data in data_list:
        corrected_date = date - timedelta(minutes=15 * data_list.index(data))
        if data_type == "production" and zone_key is not None:
            production_mix = ProductionMix(wind=data["wind"], oil=data["fossil"])
            production_breakdowns.append(
                datetime=corrected_date,
                production=production_mix,
                source=SOURCE,
                zoneKey=zone_key,
            )
        elif data_type == "consumption" and zone_key is not None:
            consumption.append(
                datetime=corrected_date,
                consumption=data["consumption"],
                source=SOURCE,
                zoneKey=zone_key,
            )
        elif data_type == "exchange":
            if zone_key1 == ZoneKey("AX") and zone_key2 == ZoneKey("SE-SE3"):
                exchanges.append(
                    datetime=corrected_date,
                    netFlow=data["sweden"] * -1,
                    source=SOURCE,
                    zoneKey=ZoneKey("AX->SE-SE3"),
                )
            elif zone_key1 == ZoneKey("AX") and zone_key2 == ZoneKey("FI"):
                exchanges.append(
                    datetime=corrected_date,
                    netFlow=round(data["alink"] + data["gustavs"], 2) * -1,
                    source=SOURCE,
                    zoneKey=ZoneKey("AX->FI"),
                )
            else:
                raise ParserException(
                    "AX.py",
                    "This parser can only fetch data between Åland <-> Sweden and Åland <-> Finland",
                )
        else:
            raise ParserException(
                "AX.py",
                "The datasource currently implemented is only for production, consumption and exchange data",
                zone_key,
            )

    if data_type == "production":
        return production_breakdowns.to_list()
    elif data_type == "consumption":
        return consumption.to_list()
    elif data_type == "exchange":
        return exchanges.to_list()
    else:
        raise ParserException(
            "AX.py",
            "The datasource currently implemented is only for production, consumption and exchange data",
            zone_key,
        )


def fetch_production(
    zone_key: ZoneKey = ZoneKey("AX"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch production data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=zone_key,
        zone_key1=None,
        zone_key2=None,
        session=session,
        logger=logger,
        data_type="production",
    )


def fetch_consumption(
    zone_key: ZoneKey = ZoneKey("AX"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch consumption data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=zone_key,
        zone_key1=None,
        zone_key2=None,
        session=session,
        logger=logger,
        data_type="consumption",
    )


def fetch_exchange(
    zone_key1: ZoneKey,
    zone_key2: ZoneKey,
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch exchange data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=None,
        zone_key1=zone_key1,
        zone_key2=zone_key2,
        session=session,
        logger=logger,
        data_type="exchange",
    )
