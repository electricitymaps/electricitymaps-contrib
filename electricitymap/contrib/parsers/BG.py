from datetime import datetime, timezone
from logging import Logger, getLogger

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

PARSER = "BG.py"

SOURCE = "eso.bg"
SOURCE_API_URL = "http://www.eso.bg/api/rabota_na_EEC_json.php?en"

PRODUCTION_TYPE_TO_PRODUCTION_MODE = {
    "NPP": "nuclear",
    "CHP": "coal",
    "Heating TPPs": "gas",
    "Factory TPPs": "gas",
    "HPP": "hydro",
    "Small HPPs": "hydro",
    "Wind": "wind",
    "PV": "solar",
    "Bio": "biomass",
}


def fetch_production(
    zone_key: ZoneKey = ZoneKey("BG"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    session = session or Session()

    if target_datetime is not None:
        raise ParserException(
            PARSER, "This parser is not yet able to parse historical data", zone_key
        )

    time = datetime.now(timezone.utc)
    response = session.get(SOURCE_API_URL)
    if not response.ok:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            zone_key,
        )

    response_payload = response.json()
    logger.debug(f"Raw generation breakdown: {response_payload}")

    production_mix = ProductionMix()
    for header, value in response_payload:
        production_type, _delimiter, _percentage = header.rpartition(" ")
        production_mode = PRODUCTION_TYPE_TO_PRODUCTION_MODE[production_type]
        production_mix.add_value(production_mode, value)

    production_breakdown_list = ProductionBreakdownList(logger)
    production_breakdown_list.append(
        zoneKey=zone_key,
        datetime=time,
        source=SOURCE,
        production=production_mix,
    )
    return production_breakdown_list.to_list()
