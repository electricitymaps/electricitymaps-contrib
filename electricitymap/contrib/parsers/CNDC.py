#!/usr/bin/env python3

import json
import re
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    EventSourceType,
    ProductionBreakdownList,
    ProductionMix,
    TotalProductionList,
)
from electricitymap.contrib.lib.types import ZoneKey

tz_bo = ZoneInfo("America/La_Paz")

INDEX_URL = "https://www.cndc.bo/gene/index.php"
DATA_URL = "https://www.cndc.bo/gene/dat/gene.php?fechag={0}"
SOURCE = "cndc.bo"


def extract_xsrf_token(html):
    """Extracts XSRF token from the source code of the generation graph page."""
    return re.search(r'var ttoken = "([a-f0-9]+)";', html).group(1)


def get_datetime(query_date: datetime, hour: int) -> datetime:
    return datetime(
        year=query_date.year,
        month=query_date.month,
        day=query_date.day,
        hour=hour,
        tzinfo=tz_bo,
    )


def fetch_data(
    session: Session | None = None, target_datetime: datetime | None = None
) -> tuple[list[dict], datetime]:
    if session is None:
        session = Session()

    if target_datetime is None:
        target_datetime = datetime.now()
    target_datetime = target_datetime.astimezone(tz_bo)
    # Define actual and previous day (for midnight data).
    formatted_dt = target_datetime.strftime("%Y-%m-%d")

    # XSRF token for the initial request
    xsrf_token = extract_xsrf_token(session.get(INDEX_URL).text)

    resp = session.get(
        DATA_URL.format(formatted_dt), headers={"x-csrf-token": xsrf_token}
    )

    hour_rows = json.loads(resp.text.replace("ï»¿", ""))["data"]
    return hour_rows, target_datetime


def parse_generation_forecast(
    zone_key: ZoneKey, date: datetime, raw_data: list[dict], logger: Logger
) -> TotalProductionList:
    result = TotalProductionList(logger)
    assert date.tzinfo == tz_bo
    for hour_row in raw_data:
        [hour, forecast, total, thermo, hydro, wind, solar, bagasse] = hour_row

        # "hour" is one-indexed
        timestamp = get_datetime(query_date=date, hour=hour - 1)

        result.append(
            zoneKey=zone_key,
            datetime=timestamp,
            value=forecast,
            source=SOURCE,
            sourceType=EventSourceType.forecasted,
        )

    return result


def parser_production_breakdown(
    zone_key: ZoneKey, date: datetime, raw_data: list[dict], logger: Logger
) -> ProductionBreakdownList:
    result = ProductionBreakdownList(logger)
    assert date.tzinfo == tz_bo
    for hour_row in raw_data:
        [hour, forecast, total, thermo, hydro, wind, solar, bagasse] = hour_row

        # "hour" is one-indexed
        timestamp = get_datetime(query_date=date, hour=hour - 1)
        modes_extracted = [hydro, solar, wind, bagasse]

        if total is None or None in modes_extracted:
            continue

        result.append(
            zoneKey=zone_key,
            datetime=timestamp,
            production=ProductionMix(
                hydro=hydro,
                solar=solar,
                wind=wind,
                biomass=bagasse,
                # NOTE: thermo includes gas + oil mixed, so we set these as unknown for now
                # The modes here should match the ones we extract in the production payload
                unknown=total - (hydro + solar + wind + bagasse),
            ),
            source=SOURCE,
        )

    return result


def fetch_production(
    zone_key: ZoneKey = ZoneKey("BO"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list:
    """Requests the last known production mix (in MW) of a given country."""
    production = ProductionBreakdownList(logger)
    raw_data, query_date = fetch_data(session=session, target_datetime=target_datetime)
    production = parser_production_breakdown(zone_key, query_date, raw_data, logger)

    return production.to_list()


def fetch_generation_forecast(
    zone_key: ZoneKey = ZoneKey("BO"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list:
    raw_data, query_date = fetch_data(session=session, target_datetime=target_datetime)
    generation_forecast = parse_generation_forecast(
        zone_key, query_date, raw_data, logger
    )
    return generation_forecast.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""
    print("fetch_production() ->")
    print(fetch_production())

    print("fetch_generation_forecast() ->")
    print(fetch_generation_forecast())
