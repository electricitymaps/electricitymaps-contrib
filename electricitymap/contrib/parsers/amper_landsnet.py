from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

SOURCE = "amper.landsnet.is"
SOURCE_URL = "https://amper.landsnet.is/generation/api/Values"


def fetch_production(
    zone_key: ZoneKey = ZoneKey("IS"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    r = session or Session()
    if target_datetime is not None:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    res = r.get(SOURCE_URL)
    if not res.ok:
        raise ParserException(
            parser="amper_landsnet.py",
            message=f"Failed to fetch {SOURCE}: {res.status_code}, err: {res.text}",
            zone_key=zone_key,
        )

    obj = res.json()
    mix = ProductionMix(
        geothermal=obj["geothermal"],
        hydro=obj["hydro"],
        oil=obj["oil"],
    )
    breakdowns = ProductionBreakdownList(logger=logger)
    breakdowns.append(
        zoneKey=zone_key,
        datetime=datetime.fromisoformat(obj["timestamp"]).replace(
            tzinfo=ZoneInfo("Atlantic/Reykjavik")
        ),
        production=mix,
        source=SOURCE,
    )
    return breakdowns.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
