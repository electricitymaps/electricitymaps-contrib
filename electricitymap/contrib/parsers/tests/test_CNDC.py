import json
from datetime import datetime
from importlib import resources

import pytest
from requests_mock import GET

from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.CNDC import (
    DATA_URL,
    INDEX_URL,
    fetch_generation_forecast,
    fetch_production,
    tz_bo,
)


@pytest.fixture(autouse=True)
def target_datetime():
    yield datetime(2023, 12, 20, tzinfo=tz_bo)


@pytest.fixture(autouse=True)
def mock_response(adapter, target_datetime):
    adapter.register_uri(
        GET,
        INDEX_URL,
        text=resources.files("electricitymap.contrib.parsers.tests.mocks.CNDC")
        .joinpath("index.html")
        .read_text(),
    )

    adapter.register_uri(
        GET,
        DATA_URL.format(target_datetime.strftime("%Y-%m-%d")),
        json=json.loads(
            resources.files("electricitymap.contrib.parsers.tests.mocks.CNDC")
            .joinpath("data.json")
            .read_text()
        ),
    )


def test_fetch_generation_forecast(session, snapshot, target_datetime):
    assert snapshot == fetch_generation_forecast(
        ZoneKey("BO"),
        session=session,
        target_datetime=target_datetime,
    )


def test_fetch_production(session, snapshot, target_datetime):
    assert snapshot == fetch_production(
        ZoneKey("BO"),
        session=session,
        target_datetime=target_datetime,
    )
