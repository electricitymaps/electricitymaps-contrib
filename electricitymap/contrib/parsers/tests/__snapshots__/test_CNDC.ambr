# serializer version: 1
# name: test_fetch_generation_forecast
  list([
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1199.23,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1141.49,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1108.31,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 3, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1088.02,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 4, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1118.32,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 5, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1080.45,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 6, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1228.59,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 7, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1292.94,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 8, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1399.9,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 9, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1454.64,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 10, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1485.06,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 11, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1499.31,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1488.93,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 13, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1547.23,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 14, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1590.45,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 15, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1572.62,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 16, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1516.26,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 17, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1442.56,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 18, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1571.34,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 19, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1627.02,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 20, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1599.87,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 21, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1525.8,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 22, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1411.17,
      'zoneKey': 'BO',
    }),
    dict({
      'datetime': datetime.datetime(2023, 12, 20, 23, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'value': 1294.64,
      'zoneKey': 'BO',
    }),
  ])
# ---
# name: test_fetch_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 384.86,
        'solar': 0,
        'unknown': 744.75,
        'wind': 45.39,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 402.78,
        'solar': 0,
        'unknown': 696.4,
        'wind': 47.16,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 406.37,
        'solar': 0,
        'unknown': 661.4,
        'wind': 48.76,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 3, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 380.18,
        'solar': 0,
        'unknown': 662.76,
        'wind': 47.33,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 4, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 379.26,
        'solar': 0,
        'unknown': 661.15,
        'wind': 51.32,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 5, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 350.42,
        'solar': 0.39,
        'unknown': 647.5,
        'wind': 50.26,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 6, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 422.66,
        'solar': 13.45,
        'unknown': 670.38,
        'wind': 41.49,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 7, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 500.35,
        'solar': 47.85,
        'unknown': 717.29,
        'wind': 45.95,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 8, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 505.62,
        'solar': 90.01,
        'unknown': 769.71,
        'wind': 46.62,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 9, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 523.88,
        'solar': 113.07,
        'unknown': 845.12,
        'wind': 32.52,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 20, 10, 0, tzinfo=zoneinfo.ZoneInfo(key='America/La_Paz')),
      'production': dict({
        'biomass': 0,
        'hydro': 511.72,
        'solar': 124.97,
        'unknown': 920.99,
        'wind': 24.76,
      }),
      'source': 'cndc.bo',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'BO',
    }),
  ])
# ---
